File: LeetCode/DataStructure/112. Path Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:        
        if root is None:
            return False
        
        def hasPathSumHelp(node, running_sum, sum):
            
            running_sum += node.val

            if node.left is None and node.right is None:
                if running_sum == sum:
                    return True
                else:
                    return False
        
            if node.left:
                if hasPathSumHelp(node.left, running_sum, sum):
                    return True
                
            if node.right:
                if hasPathSumHelp(node.right, running_sum, sum):
                    return True
            
        if hasPathSumHelp(root, 0, targetSum):
            return True
        
        return False

File: LeetCode/DataStructure/112. Path Sum.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool HasPathSum(TreeNode root, int targetSum) {
        if (root == null)
            return false;
        
        if (HasPathSumHelp(root, 0, targetSum))
            return true;
            
        return false;
    }
    
    public bool HasPathSumHelp(TreeNode node, int curSum, int targetSum)
    {
        curSum += node.val;
        if (node.left == null && node.right == null)
        {
            if (curSum == targetSum)
                return true;
            else
                return false;
        }
        
        if (node.left != null)
        {
            if (HasPathSumHelp(node.left, curSum, targetSum))
                return true;
        }
        
        if (node.right != null)
        {
            if (HasPathSumHelp(node.right, curSum, targetSum))
                return true;
        }
        
        return false;
    }
}